esteid/bdoc2/xmlsig.py

```diff
@@ -0,0 +1,345 @@
+        try:
+            c14n_alg = sp_ref_node.find('.//ds:Transform').attrib['Algorithm']
+        except:  # noqa: E722
+            c14n_alg = None
+
+        signed_props_node = self._get_node('xades:SignedProperties')
+        time_node = signed_props_node.find('.//xades:SigningTime', namespaces=self.NAMESPACES)
+        # Add a UTC timestamp. Can't use isoformat() as it adds +00:00 and microseconds
+        #  which can break the parser elsewhere
+        time_node.text = get_utc_time().strftime('%Y-%m-%dT%H:%M:%SZ')
+
+        signed_props_c14n = self.canonicalize(signed_props_node, c14n_alg)
+        # TODO select algorithm based on DigestMethod // update DigestMethod
+        signed_props_hash = hashlib.sha256(signed_props_c14n).digest()
+        sp_ref_node.find('.//ds:DigestValue', namespaces=self.NAMESPACES).text = base64.b64encode(signed_props_hash)
+
+        return self
+
+    def signed_data(self):
+        sign_info_node = self._get_node('ds:SignedInfo')
+        return self.canonicalize(sign_info_node)
+
+    def digest(self):
+        return hashlib.sha256(self.signed_data()).digest()
+
+    def get_signature_value(self):
+        sig_value_node = self._get_node('ds:SignatureValue')
+        return base64.b64decode(sig_value_node.text)
+
+    def add_signature_value(self, signature):
+        """Insert the base64-encoded value of a signature obtained from a signing service
+
+        NOTE: the signature method should be known in advance, as it's part of the SignedInfo structure over which
+          the signature is calculated.
+
+        :param signature: Binary signature
+        :return:
+        """
+        sig_value_node = self._get_node('ds:SignatureValue')
+        sig_value_node.text = base64.b64encode(signature)
+        return self
+
+    def add_root_ca_cert(self, cert_file):
+        """Add a root CA cert from file.
+
+        :param cert_file: can be one of self.ROOT_CA_CERT, self.TEST_ROOT_CA_CERT, or an absolute path
+        """
+        certs_node = self._get_node('xades:CertificateValues')
+        ca_node = etree.Element('{%s}EncapsulatedX509Certificate' % self.NAMESPACES['xades'])
+        ca_node.attrib['Id'] = (u'%s-ROOT-CA-CERT' % self.NEW_SIGNATURE_ID).encode('ascii')
+        with open(os.path.join(certs.CERT_PATH, cert_file), 'rb') as f:
+            root_cert = load_certificate(f.read())
+        ca_node.text = base64.b64encode(root_cert.asn1.dump())
+        certs_node.append(ca_node)
+        return self
+
+    def add_ocsp_response(self, ocsp_response):
+        """
+        Embed the OCSP response and certificates
+
+        :param OCSP ocsp_response:
+        :return: self
+        """
+        ocsp_response_node = self._get_node('xades:EncapsulatedOCSPValue')
+        ocsp_response_node.text = base64.b64encode(ocsp_response.get_encapsulated_response())
+        #
+        # ocsp_certs_node = self._get_node('xades:CertificateValues')
```

Comment:
Any reason to keep this? If there is, I'll add a comment about it.

Reply:
The OCSP certificate is usually embedded in a response, if it's not, then it's required to add it explicitly into the XML, this code would do it. I could probably add a switch that would control this behavior in case one uses an OCSP service which might not include its cert in a response.